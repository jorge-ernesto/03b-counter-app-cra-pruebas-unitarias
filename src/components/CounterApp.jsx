// Importar librerias
import React from 'react';
import PropTypes from 'prop-types';
import { useState } from 'react';

export const CounterApp = ({ value }) => {

    /****************** Hooks ******************/

    const [counter, setCounter] = useState(value);

    /****************** Functiones ******************/

    const handleTest = (event) => {
        console.log('event', event);
    }
    const handleAdd = () => {
        setCounter(counter + 1);
    }
    const handleSubstract = () => setCounter(counter - 1);
    const handleReset = () => setCounter(value);

    /****************** Respuesta ******************/

    return (
        <>
            <h1>CounterApp</h1>
            <h2>{counter}</h2>
            <button onClick={(event) => handleTest(event)}>Test</button>
            <button onClick={handleAdd}>+1</button>
            <button onClick={handleSubstract}>-1</button>
            <button aria-label="btn-reset" onClick={handleReset}>Reset</button>
        </>
    );
}

// Validar props
CounterApp.propTypes = {
    value: PropTypes.number.isRequired
}
