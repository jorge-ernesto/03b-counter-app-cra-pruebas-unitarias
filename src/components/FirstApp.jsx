// Importar librerias
import React from 'react';
import PropTypes from 'prop-types';

// Declarar variables
const newMessage = {
    message: 'Hola Mundo',
    title: 'Fernando'
};
const getResult = () => {
    return 4 + 4;
}

export const FirstApp = ({ title = 'No hay titulo', subTitle = 'No hay subtitulo', name = 'Fernando Herrera' }) => {

    // Debug
    // console.log('debug', { title, subTitle, name });

    // Validar props
    // if (!title || typeof title !== 'string') {
    //     throw new Error('La variable title no existe');
    // }
    // if (!subTitle || typeof subTitle !== 'number') {
    //     throw new Error('La variable subTitle no existe');
    // }

    /****************** Respuesta ******************/

    return (
        <>
            <h1 data-testid="test-title">{title}</h1>
            <p>{subTitle}</p>
            <p>{subTitle}</p>
            <p>{name}</p>
            <hr />

            <h2>{getResult()}</h2>
            <hr />

            <h1>Fernando</h1>
            <code>{JSON.stringify(newMessage)}</code>
            <p>Soy un subtitulo</p>
        </>
    );
}

// Validar props
FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.number.isRequired
}
