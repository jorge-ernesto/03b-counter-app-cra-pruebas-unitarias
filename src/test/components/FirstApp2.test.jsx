import { render, screen } from '@testing-library/react';
import { FirstApp } from '../../components/FirstApp';

describe('Pruebas en <FirstApp />', () => {

    const title = "Hola, soy Goku";
    const subTitle = 777;

    test('debe de hacer match con el snapshot', () => {

        // 1. inicialización
        const { container } = render(
            <FirstApp
                title={title}
                subTitle={123}
            />
        );

        // 3. observar el comportamiento... esperado
        expect(container).toMatchSnapshot();

    });

    test('debe de mostrar el mensaje "Hola, soy Goku"', () => {

        // 1. inicialización
        render(
            <FirstApp
                title={title}
                subTitle={123}
            />
        );

        // Debug
        // console.log('screen', screen); // Objeto screen del ultimo render
        // screen.debug(); // HTML del ultimo render

        // 3. observar el comportamiento... esperado
        expect(
            screen.getByText(title)
        ).toBeTruthy();

    });

    test('debe de mostrar el título en un h1', () => {

        // 1. inicialización
        render(
            <FirstApp
                title={title}
                subTitle={123}
            />
        );

        // Debug
        // level 1 significa h1
        console.log(
            'heading',
            screen.getAllByRole('heading', { level: 1 })[0].innerHTML
        );

        // 3. observar el comportamiento... esperado
        expect(
            screen.getAllByRole('heading', { level: 1 })[0].innerHTML
        ).toContain(title);

    });

    test('debe de mostrar el subtitulo enviado por props', () => {

        // 1. inicialización
        render(
            <FirstApp
                title={title}
                subTitle={subTitle}
            />
        );

        // 3. observar el comportamiento... esperado
        // Comprobar texto
        expect(
            screen.getAllByText(subTitle).length
        ).toBe(2);

    });

});
