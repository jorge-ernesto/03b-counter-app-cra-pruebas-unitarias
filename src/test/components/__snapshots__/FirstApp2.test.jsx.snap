// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <FirstApp /> debe de hacer match con el snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
    Hola, soy Goku
  </h1>
  <p>
    123
  </p>
  <p>
    123
  </p>
  <p>
    Fernando Herrera
  </p>
  <hr />
  <h2>
    8
  </h2>
  <hr />
  <h1>
    Fernando
  </h1>
  <code>
    {"message":"Hola Mundo","title":"Fernando"}
  </code>
  <p>
    Soy un subtitulo
  </p>
</div>
`;
